proc/groupby.go

```diff
@@ -47,8 +47,32 @@ const defaultGroupByLimit = 1000000
 
 func CompileGroupBy(node *ast.GroupByProc, zctx *resolver.Context) (*GroupByParams, error) {
 	keys := make([]GroupByKey, 0)
+	astKeys := node.Keys
 	var targetNames []string
-	for _, astKey := range node.Keys {
+
+	if node.Sorted != 0 && node.Duration.Seconds == 0 && len(astKeys) > 0 {
+		if _, ok := astKeys[0].Expr.(ast.FieldExpr); !ok {
+			return nil, fmt.Errorf("compiling groupby: cannot use -sorted in conjunction with a computed primary key")
+		}
+	}
+	if node.Duration.Seconds > 0 {
+		everyKey := ast.Assignment{
+			Target: "ts",
+			Expr: &ast.FunctionCall{
+				Node:     ast.Node{"FunctionCall"},
+				Function: "Time.trunc",
+				Args: []ast.Expression{
+					&ast.FieldRead{ast.Node{"FieldRead"}, "ts"},
+					&ast.Literal{ast.Node{"Literal"}, "int64", strconv.Itoa(node.Duration.Seconds)},
+				},
+			},
+		}
+		astKeys = append([]ast.Assignment{everyKey}, astKeys...)
+	}
+	if len(astKeys) == 0 {
+		return nil, fmt.Errorf("compiling groupby: must have at least one key (or -every)")
+	}
+	for _, astKey := range astKeys {
```

Comment:
Please don't add 25 lines of unrelated code between the declarations of keys and targetNames and their uses.